{"cells":[{"cell_type":"markdown","id":"6ead2fd2-c4c5-4f2f-9de6-a1a975c1011c","metadata":{"tags":[],"id":"6ead2fd2-c4c5-4f2f-9de6-a1a975c1011c"},"source":["<div style=\"text-align:center;\">\n","  <img src=\"custom/molssi_main_horizontal.png\" style=\"display: block; margin: 0 auto; max-height:200px;\">\n","</div>\n","\n","\n","# Digital Representation of Molecules\n","\n","<strong>Author(s):</strong> Jessica A. Nash, The Molecular Sciences Software Institute\n","\n","<div class=\"alert alert-block alert-info\">\n","<h2>Overview</h2>\n","\n","<strong>Questions:</strong>\n","\n","* How are molecules represented on computers?\n","\n","* What is a SMILES?\n","    \n","* What are common molecular file formats?\n","\n","<strong>Objectives:</strong>\n","\n","* Convert molecules from chemical formula and structures to SMILES strings.\n","\n","</div>\n","\n","In order to work with molecules in a programmatically, we have to be able to represent them on computers.\n","This can be acheived many ways.\n","One way is through something called a SMILES string.\n","There are also file formats that represent molecules.\n","Examples of file formats include ``pdb``, ``mol``, ``mol2``, ``cif``, and ``mmcif``."]},{"cell_type":"markdown","id":"b8eb3cf1-79d9-40f6-a739-92b0ba19539e","metadata":{"id":"b8eb3cf1-79d9-40f6-a739-92b0ba19539e"},"source":["## Molecular Representations of Biological Molecules for Computing\n","\n","In bioinformatics, if you work with genomics you are accustomed to working with DNA sequences. For example, this is a partial DNA sequence for human dihdydrofolate reductase, an important enzyme in nucelotide metabolism.\n","\n","```\n","GAATTCATGAAAACGTAGCTCGTCCTCAAAAAAAACAGAAGAGGAGTAATCATTTTAAGGGAGAAATATA\n","TACGAAAGGAACAAGATTTTGAAGCACCCAAGCTGCCACCTACATTAAAACACGGTAGGTGGCTAAACAC\n","CAGTCTTCAATGCCCTTCCACAGCCTCAGTCTGAAAAATACTGTGCAGGTGACCCAAGTGAGGGGTCACC\n","CTTGGGCTTTTCCTGTGGCAGTATCTCTGGTTTAAAAACAAACAAACGTACTTATTGCGTTGAAGGACGG\n","CAACAGGAAGGACTCCATGATTAGTCACATCTATACCATCCTAAGAAACTTTATCCACCCAAACTGTATT\n","TCAGACTTTATAATCTAAACTACAAAAAGTGTTCACTGGGGAACTGCACAATATGACTGCTTTTAACCGT\n","```\n","\n","The DNA sequence shown here is a simplified representation of a very complex 3D structure that is part of a chromosome, an enormously complex structure. The sequence that represents this gene can be used as a string in coding - computation with strings is orders of magnitude faster than computation with 3D structures. And we can still learn a great deal about this gene simply by exploring the sequence. Likewise we can represent the RNA transcribed from this sequence as a list of characters where T is replaced by U.\n","\n","If you study proteins or proteomics, you know that protein function depends on protein structure. Protein structures involve 20 (or more) building blocks so the sequences are more complex, but the principle of representing the protein as a simple string for ease with computing still applies. Here is the sequence of the dihydrofolate reductase protein that is coded in this gene sequence above.\n","\n","```\n","MVGSLNCIVAVSQNMGIGKNGDLPWPPLRNEFRYFQRMTTTSSVEGKQNLVIMGKKTWFSIPEKNRPLKG\n","RINLVLSRELKEPPQGAHFLSRSLDDALKLTEQPELANKVDMVWIVGGSSVYKEAMNHPGHLKLFVTRIM\n","QDFESDTFFPEIDLEKYKLLPEYPGVLSDVQEEKGIKYKFEVYEKND\n","```\n","\n","As we move into cheminformatics, we often want to convert small molecule structures, like the aspirin shown here, into strings for computing ease.\n","\n","![aspirin](images/aspirin.png)\n","\n","There are three well-known string conversions for small molecules, SMILES, InChI, and InChI Key. Here are the SMILES, InChI, and InChI Key strings for aspirin.\n","\n","SMILES: CC(=O)OC1=CC=CC=C1C(=O)O\n","\n","InChI: 1S/C9H8O4/c1-6(10)13-8-5-3-2-4-7(8)9(11)12/h2-5H,1H3,(H,11,12)\n","\n","InChI Key: BSYNRYMUTXBXSQ-UHFFFAOYSA-N\n","\n","In this workshop we will use SMILES strings. SMILES syntax is explained below."]},{"cell_type":"markdown","id":"1a20ce5f-0392-40be-a64f-006929d49a5a","metadata":{"id":"1a20ce5f-0392-40be-a64f-006929d49a5a"},"source":["## Simplified Molecular-Input Line-entry System: SMILES\n","\n","SMILES stands for \"Simplified Molecular-Input Line-Entry System\" and is a way to represent molecules as a string of characters.\n","\n","Consider the molecule ethanol. The image below shows a representation that we are used to seeing in chemistry:\n","\n","![ethanol](images/ethanol.png)\n","\n","However, the SMILES representation of this molecule would be \"CCO\".\n","\n","You can read more about SMILES at [this tutorial](https://archive.epa.gov/med/med_archive_03/web/html/smiles.html), but rules for atoms and bonds are also repeated below.\n","\n","### Atoms\n","SMILES supports all elements in the periodic table. An atom is represented using its respective atomic symbol. Upper case letters refer to non-aromatic atoms; lower case letters refer to aromatic atoms. If the atomic symbol has more than one letter the second letter must be lower case.\n","\n","### Bonds\n","```\n","-\tSingle bond\n","=\tDouble bond\n","#\tTriple bond\n","*\tAromatic bond\n",".\tDisconnected structures\n","```\n","Single bonds are the default and therefore need not be entered. For example, 'CC' would mean that there is a non-aromatic carbon attached to another non-aromatic carbon by a single bond, and the computer would identify the structure as the chemical ethane. It is also assumed that the bond between two lower case atom symbols is aromatic. A blank terminates the SMILES string.\n","\n","### Branches\n","\n","A branch from a chain is specified by placing the SMILES symbol(s) for the branch between parenthesis. Some examples:\n","\n","```\n","\n","CC(O)C\t2-Propanol\n","CC(=O)C\t2-Propanone\n","```\n","\n","### Rings\n","\n","A ring is specified by placing a number directly after the SMILES symbol where the ring closure occurs. This number acts as a marker, indicating that the atoms with the same number are connected, thus forming a ring. For instance:\n","\n","```\n","C1CCCC1   cyclopentane\n","n1ccccc1  Pyridine\n","```\n","\n","### SMILES Examples\n","\n","<div style=\"text-align:center;\">\n","  <img src=\"images/smiles_example_1.png\" style=\"display: block; margin: 0 auto; max-height:300px;\">\n","</div>\n","\n","<div style=\"text-align:center;\">\n","  <img src=\"images/smiles_example_2.png\" style=\"display: block; margin: 0 auto; max-height:300px;\">\n","</div>\n","\n","### Using Online Resources\n","Most of the time, you will not need to write a SMILES string by hand. You will be able to look up a molecule's SMILES string from a web database like [PubChem](https://pubchem.ncbi.nlm.nih.gov/).\n","\n","You can also use tools like this [molecule sketcher from the Protein Data Bank](https://www.rcsb.org/chemical-sketch)\n","to draw molecules and get their SMILES strings."]},{"cell_type":"markdown","id":"b408dc12-f3c0-4777-ae00-b77e7555f068","metadata":{"id":"b408dc12-f3c0-4777-ae00-b77e7555f068"},"source":["<div class=\"alert alert-block alert-warning\">\n","<h3>Exercise</h3>\n","\n","<p> Based on what you've learned about SMILES strings, answer the following questions:\n","<p>\n","    <ul>\n","        <li> What is the SMILES for ethanol?</li>\n","        <li> What is the SMILES for water?</li>\n","        <li> What is the SMILES for benzene?</li>\n","        <li> What molecule is represented by the SMILES O=C=O?</li>\n","        <li> What is the SMILES for an amide group?<//li>\n","        <li> What is the SMILES for a 4 membered carbon chain with an amide group on the second carbon?</li>\n","    </ul>\n","</p>\n","<p>Check your answers from this previous exercise using the PDB molecule sketcher. Notice that you can type in a SMILES string and have the sketcher draw the molecule for you.</p>\n","\n","</div>"]},{"cell_type":"code","execution_count":null,"id":"03b26b3b-d8fb-4add-94aa-b48a59052a78","metadata":{"id":"03b26b3b-d8fb-4add-94aa-b48a59052a78"},"outputs":[],"source":["# Fill in your answers here:\n","# 1.\n","# 2.\n","# 3.\n","# 4.\n","# 5.\n","# 6."]},{"cell_type":"markdown","id":"4842145a-676e-4f1f-9b55-8b8296a81301","metadata":{"id":"4842145a-676e-4f1f-9b55-8b8296a81301"},"source":["## Molecular File Formats\n","\n","Molecules can also be represented using a number of different file formats. As you work more in chemistry, you may see a number of these. Sometimes you will have to pick a file format based on the software you are using or the molecular information you want to save.\n","\n","| File Format | Description                                                                 | Features                                                              | Common Uses                              |\n","|-------------|-----------------------------------------------------------------------------|-----------------------------------------------------------------------|------------------------------------------|\n","| SMILES      | Simplified Molecular Input Line Entry System                                | Line notation for representing molecular structures                   | Database               |\n","| InChI       | International Chemical Identifier                                           | Textual identifier for chemical substances                            | Databases             |\n","| MOL/SDF     | MDL MOLfile and Structure-Data File                                         | Contains 2D/3D coordinates, atoms, bonds                              | Structure visualization, cheminformatics |\n","| PDB         | Protein Data Bank format                                                    | Often used for 3D structures of proteins and nucleic acids,but can also be used for small molecules. Often does not contain molecule information, and cannot store partial charges.                           | Structural biology, bioinformatics       |\n","| XYZ         | Cartesian coordinates                                                       | Simple text format with atom types and 3D coordinates                 | Computational chemistry, molecular dynamics |     |\n","| CIF         | Crystallographic Information File                                           | Text file format for representing crystal structure data              | Crystallography                          |\n","| PQR         | Extended PDB format with partial charges and radii                          | Includes atomic coordinates, partial charges, and radii               | Electrostatics calculations              |\n","| PDBQT       | PDB format with torsion angles and charges used in AutoDock                 | Includes atomic coordinates, partial charges, torsion angles          | Molecular docking                        |\n","|MOL2   |Tripos Mol2 format|\tContains atomic coordinates, bonds, molecule types, substructures, and partial charges|\tMolecular modeling, cheminformatics, computational chemistry\n"]},{"cell_type":"markdown","id":"cb044d9f-b44c-48d8-91a6-990c0a768a5d","metadata":{"id":"cb044d9f-b44c-48d8-91a6-990c0a768a5d"},"source":["## Introduction to RDKit Molecules\n","\n","<h2>Overview</h2>\n","\n","<strong>Questions:</strong>\n","\n","* What is RDKit?\n","\n","* What is an RDKit molecule object?\n","\n","* What can I do with RDKit molecules?\n","\n","<strong>Objectives:</strong>\n","\n","* Use RDKit to create molecules in Python\n","\n","</div>\n","\n","There are Python libraries that are made for working just with chemical data. One commonly used library in Python for data science (or cheminformatics) is called [RDKit](https://en.wikipedia.org/wiki/RDKit). RDKit is an open-source cheminformatics library, primarily developed in C++ and has been under development since the year 2000. We will be using the Python interface to RDKit, though there are interfaces in other languages.\n","\n","RDKit provides a molecule object that allows you to manipulate chemical structures. It has capabilities for reading and writing molecular file formats, calculating molecular properties, and performing substructure searches. In addition, it offers a wide range of cheminformatics algorithms such as molecular fingerprint generation, similarity metrics calculation, and molecular descriptor computation. This notebook introduces RDKit basics, but we will see more of all of these topics in later notebooks.\n","\n","<div class=\"alert alert-block alert-success\">\n","<strong>Python Skills: Python Objects</strong>\n","\n","Most of this functionality is achieved through the RDKit `mol` object. In Python, we use the word \"object\" to refer to a variable type with associated data and methods.\n","One example of an object we have seen in notebooks is a list - we could also call it a \"list object\". An object has `attributes` (data) and `methods`.\n","You access information about objects with the syntax\n","```python\n","object.data\n","```\n","where data is the attribute name.\n","\n","You access object methods with the syntax\n","```python\n","object.method(arguments)\n","```\n","\n","For example, for a list \"`append` is a method that was covered in the introductory lesson.\n","\n","```\n","my_list = []\n","my_list.append(1) # \"append\" is a method\n","```\n","</div>    \n","\n","In this lesson, we will create and manipulate RDKit `mol` objects. RDKit `mol` objects represent molecules and have\n","attributes (data) and methods (actions) associated with molecules.\n","\n","We are going to use a part of RDKit called `Chem`. To use `Chem`, we first have to import it."]},{"cell_type":"code","source":["from google.colab import output\n","output.clear()"],"metadata":{"id":"Xovgh3LcqJA2"},"id":"Xovgh3LcqJA2","execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"3c5c13a5-b369-4188-8b13-356445deeb5a","metadata":{"id":"3c5c13a5-b369-4188-8b13-356445deeb5a"},"outputs":[],"source":["!pip install rdkit\n","from rdkit import Chem"]},{"cell_type":"markdown","id":"d84d6c70-fd6e-41bf-81c6-7145d6f87b0e","metadata":{"id":"d84d6c70-fd6e-41bf-81c6-7145d6f87b0e"},"source":["## Creating Molecules with RDKit\n","\n","Throughout this tutorial, it will be helpful to have access to the [RDKit documentation](https://www.rdkit.org/docs/index.html).\n","\n","To get information about molecules in RDKit, we have to first create objects representing molecules. RDKit has a molecule object that can be used to retrieve information or calculate properties. To create a molecule object, we have to communicate the molecule identity in a way that computers understand.\n","\n","We will use SMILES strings to create our objects, though RDKit also has methods for creating molecules from the file formats listed in the previous notebook.\n","\n","### Creating molecules using SMILES\n","\n","In the last lesson, we learned about molecular representations using SMILES strings. Now we will use SMILES strings to create molecule objects in RDKIT.\n","\n","We can create a representation of methane using RDKit by using the `MolFromSmiles` function in `rdkit.Chem`."]},{"cell_type":"code","execution_count":null,"id":"b9415ede-7ab5-4a35-89e8-a32d25dfe1b9","metadata":{"id":"b9415ede-7ab5-4a35-89e8-a32d25dfe1b9"},"outputs":[],"source":["methane = Chem.MolFromSmiles('C')"]},{"cell_type":"code","execution_count":null,"id":"851ae0d6-b0a2-4474-9402-5db4abf49b1c","metadata":{"id":"851ae0d6-b0a2-4474-9402-5db4abf49b1c"},"outputs":[],"source":["methane"]},{"cell_type":"markdown","id":"a20c0958-4790-4f53-bd79-deeee2733db8","metadata":{"id":"a20c0958-4790-4f53-bd79-deeee2733db8"},"source":["<div class=\"alert alert-block alert-warning\">\n","<h3>Check Your Understanding</h3>\n","<p> Create RDKit molecules for the following molecules. You can look up the SMILES strings on <a href=\"https://pubchem.ncbi.nlm.nih.gov/\">PubChem</a>:\n","<p>\n","    <ul>\n","        <li> Propane\n","        <li> Ethene\n","        <li> Cyclohexane\n","        <li> Benzene </li>\n","        <li> Acetic Acid\n","    </ul>\n","</p>\n","<p>Create variables for each molecule. The variable names should be the molecule name (all lowercase)</p>\n","</div>\n"]},{"cell_type":"code","execution_count":null,"id":"497025f0-e60a-42d1-a26e-77db46e5c495","metadata":{"id":"497025f0-e60a-42d1-a26e-77db46e5c495"},"outputs":[],"source":["# Your answers go here\n","# propane =\n","# ethene =\n","# cyclohexane =\n","# benzene =\n","acetic_acid = Chem.MolFromSmiles(\"CC(=O)O\")"]},{"cell_type":"code","execution_count":null,"id":"f87d25a4-61b2-437a-a5a6-176a2a45a813","metadata":{"id":"f87d25a4-61b2-437a-a5a6-176a2a45a813"},"outputs":[],"source":["# Try visualizing your molecules here - you visualize a molecule as putting its variable name as the last thing in a cell\n","acetic_acid"]},{"cell_type":"markdown","id":"6669ae72-6d46-4aa1-be60-b02c671fb804","metadata":{"id":"6669ae72-6d46-4aa1-be60-b02c671fb804"},"source":["## Working with RDKit Molecules\n","\n","RDKit molecule objects have a number of methods we can use to get more information about the molecule.\n","In the next few cells, we'll look at some methods that can tell us some things about the molecules we've created.\n","The next cell has settings for modifying drawing molecules. It adds labels of atom indices and sets the image size."]},{"cell_type":"code","execution_count":null,"id":"aad59d81-9142-44f2-aa5c-7e47dfc32014","metadata":{"id":"aad59d81-9142-44f2-aa5c-7e47dfc32014"},"outputs":[],"source":["from rdkit.Chem.Draw import IPythonConsole as ipc\n","from rdkit.Chem import Draw\n","\n","# Configuration for displaying in Jupyter notebooks\n","ipc.ipython_useSVG = True  # Use SVG for higher quality images\n","ipc.drawOptions.addAtomIndices = True  # Show atom indices\n","ipc.molSize = 250,250 # Set size of image"]},{"cell_type":"markdown","id":"7c062d33-fc25-4331-8a5a-adb2ed8dc440","metadata":{"id":"7c062d33-fc25-4331-8a5a-adb2ed8dc440"},"source":["<div class=\"alert alert-block alert-success\">\n","<strong>Jupyter Skills: The Tab Key</strong>\n","\n","When working with Python objects in the Jupyter notebook, you can type a variable or object name to see the methods available on that object.\n","\n","In the cell below, type `acetic_acid.` (notice how there is a dot (`.`) at the end), then press the `tab` key.\n","A list of possible methods and attributes will come up.\n","\n","Look through the methods until you find one that gives you the number of atoms in the molecule.\n","\n","</div>"]},{"cell_type":"code","execution_count":null,"id":"e1a331ad-3b4d-44e1-ad51-dc093008dacb","metadata":{"id":"e1a331ad-3b4d-44e1-ad51-dc093008dacb"},"outputs":[],"source":["# Pick a method that will tell you the number of atoms acetic acid has.\n","x = acetic_acid.GetNumAtoms()\n","print('Acetic acid has', x, 'heavy atoms.')"]},{"cell_type":"code","execution_count":null,"id":"1b7d32b7-f531-4a4f-b05e-4c6c51664d50","metadata":{"id":"1b7d32b7-f531-4a4f-b05e-4c6c51664d50"},"outputs":[],"source":["help(acetic_acid)"]},{"cell_type":"code","execution_count":null,"id":"cc30f661-dc41-4c06-82cc-a4ae762d65e7","metadata":{"id":"cc30f661-dc41-4c06-82cc-a4ae762d65e7"},"outputs":[],"source":["# Add an argument to your function to get the number of\n","# atoms that gives you the total number of atoms in acetic_acid\n","# including hydrogens\n","\n","x = acetic_acid.GetNumAtoms()\n","y = acetic_acid.GetNumAtoms(onlyExplicit = False)\n","print(f\"There are {x} heavy atoms in acetic acid and {y} total atoms in acetic acid.\")"]},{"cell_type":"markdown","id":"c88b5979-5778-444d-aa7e-b115a3a9c664","metadata":{"id":"c88b5979-5778-444d-aa7e-b115a3a9c664"},"source":["<div class=\"alert alert-block alert-info\">\n","<h2>Overview</h2>\n","\n","<strong>Questions:</strong>\n","\n","* How can I use pandas to work with molecules from rdkit?\n","\n","* How can I visualize relationships between different parts of my data?\n","\n","<strong>Objectives:</strong>\n","\n","* Review the pandas dataframe\n","  \n","* Learn to visualize structures in a pandas dataframe\n","\n","* Learn the ```apply``` tool for expanding a pandas dataframe\n","\n","</div>"]},{"cell_type":"markdown","id":"a4cc10b7-5d18-4853-bfac-a2448f722414","metadata":{"id":"a4cc10b7-5d18-4853-bfac-a2448f722414"},"source":["[Pandas](https://pandas.pydata.org/docs/) is a Python library used for data analysis and manipulation. Within the world of data science, it is a ubiquitous and widely used library. If you are learning how to analyze data in Python, it will be almost impossible to avoid pandas.\n","\n","The central data structure of pandas is called a DataFrame. Pandas DataFrames work very closely with NumPy arrays and Pandas dataframes are specifically for data which is two dimensional (rows and columns). NumPy arrays, while similar in some ways, can work with higher dimensional data.\n","\n","Pandas is very powerful. In this session, we'll be learning how to access information in pandas dataframes and how to do some basic manipulation and analysis. We are going to be looking at a dataset which gives information about the elements in the periodic table."]},{"cell_type":"code","execution_count":null,"id":"f2e82ae2-4050-4955-a205-5ed86e42cc2b","metadata":{"id":"f2e82ae2-4050-4955-a205-5ed86e42cc2b"},"outputs":[],"source":["import pandas as pd                # the pandas library\n","from rdkit.Chem import PandasTools   # rdkit functions to work with pandas\n","PandasTools.RenderImagesInAllDataFrames(images=True)    # to display chemical structures in dataframes"]},{"cell_type":"markdown","id":"6ed0f50a-7b42-49dc-acc4-f68c4cedc1a6","metadata":{"id":"6ed0f50a-7b42-49dc-acc4-f68c4cedc1a6"},"source":["## Loading a collection of small molecules\n","\n","In the next few cells, we'll read a text file that lists the SMILES strings for five nonsteroidal antiinflammatory drugs (NSAIDS) and then use them to start building a pandas dataframe. The SMILES strings for the NSAIDS can be found in ```data/nsaids.txt```."]},{"cell_type":"code","execution_count":null,"id":"6e09ef49-ad83-4f81-a0da-c90c83f02d3b","metadata":{"id":"6e09ef49-ad83-4f81-a0da-c90c83f02d3b"},"outputs":[],"source":["import os\n","os.listdir('data')\n","nsaids_file = os.path.join('data', 'nsaids.txt')\n","with open(nsaids_file,'r') as outfile:\n","    data = outfile.readlines()\n","\n","print(data)  # to make sure we have the nsaids SMILES strings\n","type(data)   # to make sure we have the ```list``` type for our data."]},{"cell_type":"code","execution_count":null,"id":"4fdc8dce-9449-4c22-b0be-f60aabd3121b","metadata":{"id":"4fdc8dce-9449-4c22-b0be-f60aabd3121b"},"outputs":[],"source":["# Start building the pandas dataframe with the SMILES strings\n","\n","nsaid_df = pd.DataFrame({'SMILES': data})\n","nsaid_df\n"]},{"cell_type":"markdown","id":"26a9b6b3-10ff-4594-b660-5b605ac30e2e","metadata":{"id":"26a9b6b3-10ff-4594-b660-5b605ac30e2e"},"source":["### The `.apply` method\n","\n","The `.apply` method in pandas is used to apply a function along a row or column of a dataframe.\n","This is useful when you have a custom function that you need to use on every value in a column, but there is not a NumPy or Pandas function for it.\n","\n","For example, we could apply the `len` function to our `Name` column to get the number of letters in the name for each element.\n","\n","```python\n","# Number of letters in name -\n","df[\"Name\"].apply(len)\n","```\n","\n","We are going to use ```apply``` to add a new column to nsaid_df that contains the rdkit molecule for each of the NSAIDs."]},{"cell_type":"code","execution_count":null,"id":"fb693a02-58a2-467d-80fa-723d61f62219","metadata":{"id":"fb693a02-58a2-467d-80fa-723d61f62219"},"outputs":[],"source":["nsaid_df[\"Molecule\"] = nsaid_df[\"SMILES\"].apply(Chem.MolFromSmiles)\n","nsaid_df.head()"]},{"cell_type":"markdown","source":["At this point, nsaid_df contains two columns: the SMILES string and the molecule. We will go one step further and use ```apply``` to add another column to the dataframe that contains a descriptor based on the rdkit molecule. Before we can do that we need to import a new sublibrary from rdkit.Chem called Descriptors. Notice that the new columns will be based on the Molecule column, which contains rdkit mol objects."],"metadata":{"id":"WqeeyIxPiF-7"},"id":"WqeeyIxPiF-7"},{"cell_type":"code","source":["from rdkit.Chem import Descriptors\n","from rdkit.Chem import Crippen\n","nsaid_df[\"Mol Wt\"] = nsaid_df[\"Molecule\"].apply(Descriptors.MolWt)\n","nsaid_df[\"logP\"] = nsaid_df[\"Molecule\"].apply(Descriptors.MolLogP)\n","nsaid_df[\"Crippen MolMR\"] = nsaid_df[\"Molecule\"].apply(Crippen.MolMR)\n","nsaid_df.head()"],"metadata":{"id":"v7lK6asEid7Q"},"id":"v7lK6asEid7Q","execution_count":null,"outputs":[]},{"source":["# @title Mol Wt vs logP\n","\n","from matplotlib import pyplot as plt\n","nsaid_df.plot(kind='scatter', x='Mol Wt', y='logP', s=32, alpha=.8)\n","plt.gca().spines[['top', 'right',]].set_visible(False)"],"cell_type":"code","execution_count":null,"outputs":[],"metadata":{"id":"2Ujik9dpmfnW"},"id":"2Ujik9dpmfnW"},{"cell_type":"code","source":["# @title Log P vs Maximum Partial Charge\n","\n","from matplotlib import pyplot as plt\n","nsaid_df.plot(kind='scatter', x='Crippen MolMR', y='logP', s=32, alpha=.8)\n","plt.gca().spines[['top', 'right',]].set_visible(False)"],"metadata":{"id":"ExINniK0kOUQ"},"id":"ExINniK0kOUQ","execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"dWKNKdDVkJxn"},"id":"dWKNKdDVkJxn","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.2"},"colab":{"provenance":[{"file_id":"1mum7oWvHmdYkO4Qrrv31pPyPxUYnepiu","timestamp":1741805025703}]}},"nbformat":4,"nbformat_minor":5}